[FIX] PepXML terminal mod mass

timosachsenberg opened the pull request:
may fix https://github.com/OpenMS/OpenMS/issues/1926

hendrikweisser commented:
Thanks, Timo. The change looks simple enough. I would just put the whole `f << ...;` statement on one line, since it's shorter now.
The "PepXMLFile_test" still needs to be updated.

I'm not sure if we'd need to make changes to the way we write the "modified_peptide" attribute in "modification_info". OpenMS currently writes something like this: "(Dimethyl)LSELGNQYSNNVLDATMGWTK(Dimethyl)"
Oliver's pepXML example (from X! Tandem?) writes something like this: "n[35]RQLNK[162]LQHK[162]GEAR"
The pepXML schema just says it should be the "Peptide sequence (with indicated modifications)", but doesn't clearly specify the format. The attribute is not marked as "required" and the information is redundant (given "peptide" in "search_hit" and the "mod_aminoacid_mass" elements), so maybe we could leave it as it is or even remove it?

timosachsenberg commented:
I have no strong opinion about this. If e.g. the pepXML encoding with brackets is the most commonly used one I would even say that we could also use it.

hroest commented:
I am pretty sure most tools (TPP etc) expect the bracket notation and they will surely break with anything else. I think it is best to stick close to what other tools expect and not rely on specifications only.

hendrikweisser commented:
What's the status here?

timosachsenberg commented:
I'll check...

hroest commented:
+1 for brackets to make it compatible with the TPP

sent from my smartphone -- you may keep any typos you find.
On Aug 2, 2016 7:05 AM, "hendrikweisser" <notifications@github.com> wrote:

> What's the status here?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/1933#issuecomment-236914641>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcUjj-Jdgv7gZqr48GkDnLvVDPhmYb_ks5qb06vgaJpZM4ID9wv>
> .
>

timosachsenberg commented:
I can do it but I might need some assistance.
Any suggestions or comments how we represent cases not supported in the TPP notation. 
e.g. we also have amino acid masses (aka mass tags and used as [ ] in various variants in our notation) and modification masses (used as ([...]) ) if I am not mistaken.
I am not too familiar with pepXML and maybe you guys have a clearer picture - otherwise I might need to dig deeper into the documentation.

hroest commented:
I think the idea of the notation is that A[+x] means that we have regular
amino acid A + mass x which represent the modification masses wherewas
unknown masses can be represented as A[x] where "x" now just represents the
mass of the amino acid. At least this is my interpretation.

How can this notation not represent all cases? Basically this allows you to
have arbitrary masses...

Hannes


On Tue, Aug 2, 2016 at 8:19 AM, Timo Sachsenberg <notifications@github.com>
wrote:

> I can do it but I might need some assistance.
> Any suggestions or comments how we represent cases not supported in the
> TPP notation.
> e.g. we also have amino acid masses (aka mass tags and used as [ ] in
> various variants in our notation) and modification masses (used as ([...])
> ) if I am not mistaken.
> I am not too familiar with pepXML and maybe you guys have a clearer
> picture - otherwise I might need to dig deeper into the documentation.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/1933#issuecomment-236938123>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcUjsGzNud0eGG-9VFAGenUvdPG0a-Eks5qb1_4gaJpZM4ID9wv>
> .
>

timosachsenberg commented:
I think (at least in the X!Tandem case) they only use nominal masses.
In addition, we support constructs like [x]([+y]) (AA mass and modification) in our string.
Does anyone know if it would break anything on TPP side if I would use exact masses [x] instead of [(int)(x)]?

hroest commented:
Maybe we should contact them?  I don't think guessing will help here. ...

sent from my smartphone -- you may keep any typos you find.
On Aug 3, 2016 1:02 AM, "Timo Sachsenberg" <notifications@github.com> wrote:

> I think (at least in the X!Tandem case) they only use nominal masses.
> In addition, we support constructs like x (AA mass and modification) in
> our string.
> Does anyone know if it would break anything on TPP side if I would use
> exact masses [x] instead of [(int)(x)]?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/1933#issuecomment-237170400>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAcUjgQArh8uoknMgkSbOpqlP14o_8v4ks5qcEsugaJpZM4ID9wv>
> .
>

timosachsenberg commented:
yep I will ask them

timosachsenberg commented:
any hints or comments on a possible bug that existed before :

(Glu->pyro-Glu) has a delta mass of -18.010565 but the actual mod_nterm_mass should be + 1 Hydrogen in my view.
So I would guess that mod_nterm_mass is also wrongly calculated.

```
before:
<modification_info modified_peptide="(Glu->pyro-Glu)ELNKEMAAEKAKAAAG" mod_nterm_mass="-18.010565">
now (using the new bracket annotation):
<modification_info modified_peptide="n[-17]ELNKEMAAEKAKAAAG" mod_nterm_mass="-18.010565">
```

timosachsenberg commented:
from other pep.xml example we have e.g.:
```
<search_hit hit_rank="1" peptide="WETEERPR" peptide_prev_aa="G" peptide_next_aa="T" protein="sp|Q13162|PRDX4_HUMAN" protein_descr="Peroxiredoxin-4 OS=Homo sapiens GN=PRDX4 PE=1 SV=1" num_tot_proteins="1" num_matched_ions="7" tot_num_ions="28" calc_neutral_pep_mass="1133.5771" massdiff="-0.004" num_tol_term="1" num_missed_cleavages="0" is_rejected="0">
<modification_info mod_nterm_mass="33.0643" mod_cterm_mass="17.0031" modified_peptide="WETEERPR"/>

and

<search_hit hit_rank="1" peptide="THNDIIHNENMR" peptide_prev_aa="R" peptide_next_aa="Q" protein="sp|P15311|EZRI_HUMAN" protein_descr="Ezrin OS=Homo sapiens GN=EZR PE=1 SV=4" num_tot_proteins="1" num_matched_ions="11" tot_num_ions="66" calc_neutral_pep_mass="1528.7597" massdiff="0.011" num_tol_term="2" num_missed_cleavages="0" is_rejected="0">
<modification_info mod_nterm_mass="37.0836" mod_cterm_mass="17.0031" modified_peptide="n[37]THNDIIHNENMR"/>
```

lars20070 commented:
@oliverschilling Can you please commend regarding #1926? Thanks.

timosachsenberg commented:
maybe @oliverschilling can provide some additional information on:
- how mod_nterm_mass and mod_cterm_mass are calculated in pepXML
- the modifications detected in the "WETEERPR" and "THNDIIHNENMR" peptide
- how these are annotated (e.g. we have "WETEERPR" and "n[37]THNDIIHNENMR")

My first guess would be that variable and fixed mods are treated differently.
See the modification definitions:
```
<aminoacid_modification aminoacid="C" massdiff="57.0215" mass="160.0306" variable="N"/>
<aminoacid_modification aminoacid="K" massdiff="32.0564" mass="160.1514" variable="N"/>
<aminoacid_modification aminoacid="C" massdiff="-17.0265" mass="143.0041" variable="Y" symbol="^"/>
<!--X! Tandem n-terminal AA variable modification-->
<aminoacid_modification aminoacid="C" massdiff="57.0215" mass="217.0521" variable="Y"/>
<aminoacid_modification aminoacid="E" massdiff="-18.0106" mass="111.0320" variable="Y" symbol="^"/>
<!--X! Tandem n-terminal AA variable modification-->
<aminoacid_modification aminoacid="K" massdiff="36.0757" mass="196.2270" variable="Y"/>
<aminoacid_modification aminoacid="Q" massdiff="-17.0265" mass="111.0321" variable="Y" symbol="^"/>
<!--X! Tandem n-terminal AA variable modification-->
<terminal_modification terminus="n" massdiff="32.0565" mass="33.0643" protein_terminus="N" variable="N"/>
<terminal_modification terminus="c" massdiff="0.0003" mass="17.0031" protein_terminus="N" variable="N"/>
<terminal_modification terminus="n" massdiff="42.0106" mass="75.0749" protein_terminus="N" variable="Y" symbol="^"/>
<!--X! Tandem n-terminal AA variable modification-->
<terminal_modification terminus="n" massdiff="36.0757" mass="69.1400" protein_terminus="N" variable="Y"/>
```
It seems as if fixed modifications are not annotated in the peptide sequence.

timosachsenberg commented:
Given the modification defintion above. Where does the 164 modification come from?
see e.g.
```
<modification_info mod_nterm_mass="37.0836" mod_cterm_mass="17.0031" modified_peptide="n[37]APLDLDK[164]YVEIAR">
<mod_aminoacid_mass position="7" mass="164.1706"/>
</modification_info>
```

timosachsenberg commented:
Summary of our discussion on the phone:
- TPP tools / converters are inconsistent how modifications are handled
- best way is probably to report total terminus masses (including modifications, e.g. H+mod and OH+mod)
- only nominal masses are reported in the sequence

I implemented these changes (fixes https://github.com/OpenMS/OpenMS/issues/1926)

At least for the xinteract, prophet, xtandem combination only variable modifications seem to be reported as nominal masses in brackets in the sequence. I don't see any important advantage in doing this and I report (for the sake of a simpler implementation) all modifications. (I think, this will most likely make no difference with downstream tools as these should parse the mod_aminoacid_mass attribute anyways).

lars20070 commented:
Yes. Correct summary of our discussion.

hroest commented on src/openms/include/OpenMS/FORMAT/PepXMLFile.h:
```diff
@@ -149,6 +149,12 @@ namespace OpenMS
+        @brief create a TPP compatible string of the modified sequence using bracket notation
+        i.e.: n[35]RQLNK[162]LQHK[162]GEA
+    */  
+    static String modifiedSequenceToBracketString_(const AASequence & seq);
```
I am not sure about putting this into `PepXMLFile`, shouldnt this go into `AASequence` or a helper class?

timosachsenberg replied:
I think this is never parsed by us from pepXML files and not required to determine the mod state but I can check this again. Given that there is no real standardized format for this string (at least that is how I understood it) I would only use it for writing out pepXML and not expose it to the API.

hroest replied:
I think we do parse it. In aasequence maybe?

sent from my smartphone -- you may keep any typos you find.
On Aug 5, 2016 12:15 PM, "Timo Sachsenberg" <notifications@github.com>
wrote:

> In src/openms/include/OpenMS/FORMAT/PepXMLFile.h
> <https://github.com/OpenMS/OpenMS/pull/1933#discussion_r73745504>:
>
> > @@ -149,6 +149,12 @@ namespace OpenMS
> >      */
> >      void matchModification_(const double mass, const String& origin, String& modification_description);
> >
> > +    /**
> > +        @brief create a TPP compatible string of the modified sequence using bracket notation
> > +        i.e.: n[35]RQLNK[162]LQHK[162]GEA
> > +    */
> > +    static String modifiedSequenceToBracketString_(const AASequence & seq);
>
> I think this is never parsed by us from pepXML files and not required to
> determine the mod state but I can check this again. Given that there is no
> real standardized format for this string (at least that is how I understood
> it) I would only use it for writing out pepXML and not expose it to the
> API.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenMS/OpenMS/pull/1933/files/60d1854d6776d7a431c9ffdaadde3576bfca9db8..23c9880fd1e1b6a59b9eebfcfb5e01f40688cf7e#r73745504>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAcUjkQleQwtxOGoybMk8GpacA0JYrzzks5qc4u8gaJpZM4ID9wv>
> .
>

timosachsenberg replied:
I think the proper way to determine the modification state in pepXML should be via the other elements in modification_info. I will check this (probably) on Monday - maybe the parsing is needed for search engine specific files / incomplete pepXML.

hroest replied:
pepXML is not the only place we have to read these square notation, another format is when reading SpectraST format. Also other search engines use it, for example ProteinPilot even though they have yet another notation. Basically the whole TPP uses it, its not just pepXML and we have to support it at the moment whether we like it or not.

hroest commented:
The logic is for sure in AASequence, see https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/source/AASequence_test.cpp#L116 - therefore I argue that the conversion back to square notation should also happen in AASequence...

timosachsenberg commented:
Yes, I see this but as I pointed out the square notation has different versions (e.g. fixed mods are not written at all vs. fixed mods are written - something which we don't know at the level of AASequences) and is highly ambiguous because of the use of nominal masses. Therefore I would rather argue that even the parsing should actually not be part of AASequence but of a specific file format handler.

timosachsenberg commented:
On the other hand, I have no strong opinion about this and I can for sure add it to AASequence but ignore whether we have a fixed or variable modification.

hendrikweisser commented:
> I think the proper way to determine the modification state in pepXML should be via the other elements in modification_info and the modification definitions.

This is true for pepXML, but I think also some other search engine formats. I think it would be useful to have a version of `AASequence::fromString` that gets the lists of allowed variable/fixed modifications and uses this information to interpret the sequence string. This would also take care of the "fixed mods are [not] written" issue. (So if you already have to write this function for pepXML, you might as well make it slightly more general - if necessary - and put it in AASequence.)

By the way, I've noticed that the code for reading/writing a modified peptide in mzid-compliant format (with "location" and a CV param for the modification) seems to be replicated in several places, e.g. in OpenSWATH (TraML?). It may be worth putting this in AASequence (or as public static functions in MzIdentMLHandler?), too.

timosachsenberg commented:
Ok, I will put the toBracketString function into AASequence. If fixed modifications are passed then these are omitted.

I think smarter parsing is worth another issue.

timosachsenberg commented:
By the way, our pepXML implementation seem to be quite incomplete. So might be good to take closer a look at it: see e.g. https://github.com/OpenMS/OpenMS/issues/2096
Won't fix this in this PR though

timosachsenberg commented:
ok I add the parsing as well as we otherwise might run into problem if we re-read the pepXML

timosachsenberg commented:
any comments? merge?

timosachsenberg commented:
By the way, according to some feedback from the TPP guys they want to move away from the bracket + nominal mass notation to bracket + PSI-MOD id.

lars20070 commented:
+1

mwalzer commented:
I think this is fine, as long as the checks are passed again.
As @timosachsenberg mentioned, there are quite some unresolved issues with our implementation of pepXML - let's take it one PR at a time.
